RecoLocalMuon/GEMCSCSegment/plugins/GEMCSCSegAlgoRR.cc

```diff
@@ -0,0 +1,331 @@
+    {
+
+      // chain hits :: make a vector of TrackingRecHits
+      //               that contains the CSC and GEM rechits
+      //               and that can be given to the fitter
+      chainedRecHits = this->chainHitsToSegm(*cscSegIt, gemrechits);
+  
+      // if gemrechits are associated, run the buildSegments step
+      if(chainedRecHits.size() > (*cscSegIt)->specificRecHits().size()) 
+	{
+	  segmentvectorfromfit = this->buildSegments(*cscSegIt, chainedRecHits);
+	}
+  
+      // if no gemrechits are associated, wrap the existing CSCSegment in a GEMCSCSegment class
+      else 
+	{
+	  std::vector<const GEMRecHit*> gemRecHits_noGEMrh; // empty GEMRecHit vector
+	  GEMCSCSegment tmp(*cscSegIt, gemRecHits_noGEMrh, (*cscSegIt)->localPosition(), (*cscSegIt)->localDirection(), (*cscSegIt)->parametersError(), (*cscSegIt)->chi2());
+	  segmentvectorfromfit.push_back(tmp);
+	}
+
+      segmentvectorchamber.insert( segmentvectorchamber.end(), segmentvectorfromfit.begin(), segmentvectorfromfit.end() );
+      segmentvectorfromfit.clear();
+    }
+
+  // add the found gemcsc segments to the collection of all gemcsc segments and return
+  segmentvectorfinal.insert( segmentvectorfinal.end(), segmentvectorchamber.begin(), segmentvectorchamber.end() );  
+  segmentvectorchamber.clear();
+  edm::LogVerbatim("GEMCSCSegAlgoRR") << "[GEMCSCSegAlgoRR::run] GEMCSC Segments fitted or wrapped, size of vector = "<<segmentvectorfinal.size();
+  return segmentvectorfinal;
+}
+
+
+/**
+ * Chain hits :: make a TrackingRecHit vector containing both CSC and GEM rechits 
+ *               take the hits from the CSCSegment and add a clone of them
+ *               take the hits from the GEM RecHit vector and 
+ *               check whether they are compatible with the CSC segment extrapolation
+ *               to the GEM plane. If they are compatible, add these GEM rechits
+ */
+std::vector<const TrackingRecHit*> GEMCSCSegAlgoRR::chainHitsToSegm(const CSCSegment* cscsegment, const std::vector<const GEMRecHit*>& gemrechits) 
+{
+
+  std::vector<const TrackingRecHit*> chainedRecHits;
+
+  // working with layers makes it here a bit more difficult:
+  // the CSC segment points to the chamber, which we cannot ask from the map
+  // the CSC rechits point to the layers, from which we can ask the chamber
+
+  auto segLP                   = cscsegment->localPosition();
+  auto segLD                   = cscsegment->localDirection();
+  auto cscrhs                  = cscsegment->specificRecHits();
+
+  // Loop over the CSC Segment rechits 
+  // and save a copy in the chainedRecHits vector
+  for (auto crh = cscrhs.begin(); crh!= cscrhs.end(); crh++)
+    {
+      chainedRecHits.push_back(crh->clone());
+    }
+
+  // now ask the layer id of the first CSC rechit
+  std::vector<const TrackingRecHit*>::const_iterator trhIt = chainedRecHits.begin();
+  // make sure pointer is valid 
+  if(trhIt != chainedRecHits.end()) {
```

Comment:
Isn't the logic here backwards? It should be:
if (trhIt == chainedRecHits.end()) {

Or am I confused?